cookbook/controller/argument_value_resolver.rst

```diff
@@ -0,0 +1,234 @@
+creating and registering custom argument value resolvers, you can extend
+this functionality.
+
+Functionality Shipped With The HttpKernel
+-----------------------------------------
+
+Symfony ships with four value resolvers in the HttpKernel:
+
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\ArgumentFromAttributeResolver`
+    Attempts to find a request attribute that matches the name of the argument.
+
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\RequestValueResolver`
+    Injects the current ``Request`` if type-hinted with ``Request``, or a
+    sub-class thereof.
+
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\DefaultValueResolver`
+    Will set the default value of the argument if present and the argument
+    is optional.
+
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\VariadicValueResolver`
+    Verifies in the request if your data is an array and will add all of
+    them to the argument list. When the action is called, the last (variadic)
+    argument will contain all the values of this array.
+
+.. note::
+
+    Prior to Symfony 3.1, this logic was resolved within the ``ControllerResolver``.
+    The old functionality is rewritten to the aforementioned value resolvers.
+
+Adding a Custom Value Resolver
+------------------------------
+
+Adding a new value resolver requires one class and one service defintion.
+In the next example, you'll create a value resolver to inject the ``User``
+object from the security system. Given you write the following action::
+
+    namespace AppBundle\Controller;
+
+    use AppBundle\Entity\User;
+    use Symfony\Component\HttpFoundation\Response;
+
+    class UserController
+    {
+        public function indexAction(User $user)
+        {
+            return new Response('<html><body>Hello '.$user->getUsername().'!</body></html>');
+        }
+    }
+
+Somehow you will have to get the ``User`` object and inject it into the controller.
+This can be done by implementing the :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolverInterface`.
+This interface specifies that you have to implement two methods::
+
+    interface ArgumentValueResolverInterface
+    {
+        public function supports(Request $request, ArgumentMetadata $argument);
+        public function resolve(Request $request, ArgumentMetadata $argument);
+    }
+
+``supports()``
+    This method is used to check whether the value resolver supports the
+    given argument. ``resolve()`` will only be executed when this returns ``true``.
+``resolve()``
+    This method will resolve the actual value for the argument. Once the value
+    is resolved, you must `yield`_ the value to the ``ArgumentResolver``.
+
+Both methods get the ``Request`` object, which is the current request, and an
+:class:`Symfony\\Component\\HttpKernel\\ControllerMetadata\\ArgumentMetadata`.
```

Comment:
I would add "instance" in this sentence: "Both methods get [...] and an `ArgumentMetadata` instance"